src/solvers/petsc_dm_wrapper.C

```diff
@@ -16,39 +16,355 @@
+    PetscErrorCode
+    __libmesh_petsc_DMCreateInterpolation (DM dmc /*coarse*/, DM dmf /*fine*/,
+                                         Mat * mat ,Vec * vec)
+    {
+      libmesh_assert(dmc);
+      libmesh_assert(dmf);
+      libmesh_assert(mat);
+      libmesh_assert(vec); // Optional scaling (not needed for mg)
+
+      // Get a communicator from incoming DM
+      PetscErrorCode ierr;
+      MPI_Comm comm;
+      PetscObjectGetComm((PetscObject)dmc, &comm);
+
+      // Extract our coarse context from the incoming DM
+      void * ctx_c = NULL;
+      ierr = DMShellGetContext(dmc, &ctx_c);
+      LIBMESH_CHKERR(ierr);
+      libmesh_assert(ctx_c);
+      PetscDMContext * p_ctx_c = static_cast<PetscDMContext*>(ctx_c);
+
+      // Extract our fine context from the incoming DM
+      void * ctx_f = NULL;
+      ierr = DMShellGetContext(dmf, &ctx_f);LIBMESH_CHKERR(ierr);
+      libmesh_assert(ctx_f);
+      PetscDMContext * p_ctx_f = static_cast<PetscDMContext*>(ctx_f);
+
+      // Check for existing projection matrix
+      libmesh_assert(p_ctx_c->K_interp_ptr);
+      libmesh_assert(p_ctx_c->K_sub_interp_ptr);
+
+      // If were doing fieldsplit we need to construct sub projection
+      // matrices. We compare the passed in number of DMs fields to a
+      // global DM in order to determine if a subprojection is needed.
+      PetscInt nfieldsc,nfieldsf, nfieldsg;
+      libmesh_assert(p_ctx_c->global_dm);
+      DM * globaldm = p_ctx_c->global_dm;
+
+      ierr = DMCreateFieldIS(dmc, &nfieldsc, NULL, NULL);
+      LIBMESH_CHKERR(ierr);
+      ierr = DMCreateFieldIS(dmf, &nfieldsf, NULL, NULL);
+      LIBMESH_CHKERR(ierr);
+      ierr = DMCreateFieldIS(*globaldm, &nfieldsg, NULL, NULL);
+      LIBMESH_CHKERR(ierr);
+
+      // If subfields are identified, were doing FS so we need to create the subProjectionMatrix
+      if (nfieldsc < nfieldsg)
+        {
+          // Loop over the fields and merge their index sets.
+          std::vector<std::vector<numeric_index_type>> allrows,allcols;
+          std::vector<numeric_index_type> rows,cols;
+          allrows = p_ctx_f->dof_vec;
+          allcols = p_ctx_c->dof_vec;
+
+          // For internal libmesh submat extraction need to merge all
+          // field dofs and then sort the vectors so that they match
+          // the Projection Matrix ordering
+          static const int n_subfields = nfieldsc;
+          if ( n_subfields > 1 )
+            {
+              for (int i = 0 ; i < n_subfields ; i++)
+                {
+                  rows.insert(rows.end(), allrows[i].begin(), allrows[i].end());
+                  cols.insert(cols.end(), allcols[i].begin(), allcols[i].end());
+                }
+              std::sort(rows.begin(),rows.end());
+              std::sort(cols.begin(),cols.end());
+            }
+
+          // Now that we have merged the fine and coarse index sets
+          // were ready to make the submatrix and pass it off to PETSc
+          Mat  submat;
```

Comment:
Is `submat` actually used here? Does PETSc have ownership of it somehow, or does it need to be destroyed?

Reply:
MatCreate usually needs to be matched with MatDestroy, yeah.  This looks like a memory leak?

Reply:
Oops, left around from one of the initial attempts.